xarray/core/alignment.py

```diff
@@ -264,13 +276,14 @@ def find_matching_indexes(self) -> None:
         self.all_indexes = all_indexes
         self.all_index_vars = all_index_vars
 
-        if self.join == "override":
+        if self.join == "override" or not self.broadcast:
             for dim_sizes in all_indexes_dim_sizes.values():
                 for dim, sizes in dim_sizes.items():
                     if len(sizes) > 1:
                         raise ValueError(
-                            "cannot align objects with join='override' with matching indexes "
-                            f"along dimension {dim!r} that don't have the same size"
+                            f"cannot align objects with join={self.join!r} or "
+                            f"broadcast={self.broadcast!r} with matching indexes "
+                            f"along dimension {dim!r} that don't have the same size ({sizes!r})"
```

Comment:
Perhaps this is clearer:
```suggestion
                            f"cannot align objects"
                            f"with indexes "
                            f"along dimension {dim!r} that don't have the same size ({sizes!r}) when {message} "
```

where

```
message = "join='override'" if self.join == "override" else "broadcast=False`
```